# Лабораторная работа №1
## Выполнил студент группы БФИ2001 Писаков Андрей Андреевич

### Оглавление
1. [Задание 1](#Задание-№1)
2. [Задание 2](#Задание-№2)
3. [Задание 3](#Задание-№3)
4. [Вывод](#Вывод)

### Задание №1
<i> Вызвать функцию print() и передать туда строку Hello, World! </i>

In [1]:
print("Hello, World!")

Hello, World!


### Задание №2
Написать генератор случайных матриц(многомерных), который принимает
опциональные параметры <b>m</b>, <b>n</b>, <b>min_limit</b>, <b>max_limit</b>, где <b>m</b> и <b>n</b> указывают размер
матрицы, а <b>min_lim</b> и <b>max_lim</b> - минимальное и максимальное значение для
генерируемого числа.

In [2]:
import random
import time

In [24]:
def generateMatrix(m, n, min_lim, max_lim):
    array = [[0] * n for i in range(m)]
    for i in range(m): 
        for j in range(n): 
            array[i][j] = random.randint(min_lim, max_lim)
    return array


user_m = int(input())
user_n = int(input())
user_min_limit = int(input())
user_max_limit = int(input())
a = generateMatrix(user_m, user_n, user_min_limit, user_max_limit)

 300
 300
 -100
 100


### Задание №3
Реализовать методы сортировки строк числовой матрицы в соответствии с
заданием. Оценить время работы каждого алгоритма сортировки и сравнить его со
временем стандартной функции сортировки. Испытания проводить на сгенерированных
матрицах.

In [5]:
import copy

In [25]:
# Сортировка выбором.
# Берётся срез массива, в котором минимальный элемент переносят в самый левый угол,
# после чего срез уменьшается и цикл повторяется.
def sort1(n, array):
    for row in array:
        i = 0
        while i < n - 1:
            k = i
            j = i + 1
            while j < n:
                if (row[j] < row[k]): k = j
                j += 1
            row[i], row[k] = row[k], row[i]
            i += 1
            
            
start_time = time.time()
sort1(user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 1906 ms ---


In [26]:
# Сортировка вставкой.
def sort2(n, array):
    for row in array:
        for i in range(1, n):
            key = row[i]
            j = i - 1
            while j >= 0 and key < row[j]:
                row[j + 1] = row[j]
                j -= 1
            row[j + 1] = key
            
            
start_time = time.time()
sort2(user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 1391 ms ---


In [27]:
# Сортировка обменом. || Сортировка пузырьком.
def sort3(n, array):
    for row in array:
        for i in range(n - 1):
            for j in range(n - i - 1):
                if row[j] > row[j + 1]:
                    row[j], row[j + 1] = row[j + 1], row[j]
                    
                    
start_time = time.time()
sort3(user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 2740 ms ---


In [28]:
import math
# Сортировка Шелла.
def sort4(n, array):
    for row in array:
        k = int(math.log2(n))
        if k % 2 == 0: interval = 9 * 2 ** k - 9 * 2 ** (k // 2) + 1
        else: interval = 8 * 2 ** k - 6 * 2 ** ((k + 1) // 2) + 1
        while interval > 0:
            for i in range(interval, n):
                temp = row[i]
                j = i
                while j >= interval and row[j - interval] > temp:
                    row[j] = row[j - interval]
                    j -= interval
                row[j] = temp
            k -= 1
            if k % 2 == 0: interval = 9 * 2 ** k - 9 * 2 ** (k // 2) + 1
            else: interval = 8 * 2 ** k - 6 * 2 ** ((k + 1) // 2) + 1
            
            
start_time = time.time()
sort4(user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 271 ms ---


In [29]:
# Быстрая сортировка.
def sort6(n, array):
    for row in array:
        row_sort(n, row, 0, n - 1)
def row_sort(n, row, fst, lst):
    if fst >= lst: return
    i, j = fst, lst
    d = row[random.randint(fst, lst)]
    while i <= j:
        while row[i] < d: i += 1
        while row[j] > d: j -= 1
        if i <= j:
            row[i], row[j] = row[j], row[i]
            i, j = i + 1, j - 1
    row_sort(n, row, fst, j)
    row_sort(n, row, i, lst)
    
    
start_time = time.time()
sort6(user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000))) 

--- 365 ms ---


In [30]:
# Турнирная сортировка.
def sort5(m, n, array):
    for i in range(m):
        losers = []
        winners = []
        heap = []
        cycle(heap, winners, losers, array[i], n)
        winners2 = []
        while len(losers) != 0:
            newrow = losers.copy()
            winners2.clear()
            losers.clear()
            cycle(heap, winners2, losers, newrow, len(newrow))
            newrow.clear()
            merge(winners, winners2, newrow)
            winners = newrow.copy()
        array[i] = winners
def cycle(heap, winners, losers, row, n):
    heap_increase_key(heap, len(heap), row[0])
    if n != 1:
        heap_increase_key(heap, len(heap), row[1])
        for j in range(2, n):
            heap_insert(heap, row[j], losers, winners, len(heap))
    while len(heap) != 0:
        winners.append(heap_extract_min(len(heap), heap))
def merge(winners, winners2, newrow):
    while len(winners) != 0 and len(winners2) != 0:
        if winners[0] < winners2[0]:
            newrow.append(winners[0])
            winners.remove(winners[0])
        else:
            newrow.append(winners2[0])
            winners2.remove(winners2[0])
    if len(winners) != 0:
        newrow += winners
        winners.clear()
    else:
        newrow += winners2
        winners2.clear()
def min_heapify(heap, index, n):
    l = 2 * index + 1
    r = 2 * index + 2
    largest = l if l < n and heap[l] < heap[index] else index
    if r < n and heap[r] < heap[largest]: largest = r
    if largest != index:
        heap[largest], heap[index] = heap[index], heap[largest]
        min_heapify(heap, largest, n)
def heap_extract_min(n, heap):
    min = heap[0]
    if n != 1:
        heap[0] = math.inf
        min_heapify(heap, 0, n)
        heap.remove(math.inf)
        n -= 1
    else: heap.remove(heap[0])
    return min
def heap_increase_key(heap, index, key):
    heap.append(key)
    while index > 0 and heap[(index - 1) // 2] > heap[index]:
        heap[(index - 1) // 2], heap[index] = heap[index], heap[(index - 1) // 2]
        index = (index - 1) // 2
def heap_insert(heap, key, losers, winners, size):
    if len(winners) != 0 and winners[len(winners) - 1] > key: losers.append(key)
    elif len(winners) != 0 and winners[len(winners) - 1] == key: winners.append(key)
    else:
        heap_increase_key(heap, size, key)
        size += 1
        winners.append(heap_extract_min(size, heap))
        
        
start_time = time.time()
sort5(user_m, user_n, copy.deepcopy(a))
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 815 ms ---


In [31]:
# Встроенная функция сортировки

start_time = time.time()
array = copy.deepcopy(a)
for i in range(user_m):
    array[i].sort()
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 78 ms ---


### Вывод


#### Оценка время работы каждого алгоритма сортировки проводилась со следующими входными значениями:
    
    Количество строк матрицы: 300
    Количество столбцов матрицы: 300
    Минимальное значение элементов матрицы: -100
    Максимальное значение элементов матрицы: 100

    Результаты:
     --- 1906 ms --- время работы сортировки выбором
     --- 1391 ms --- время работы сортировки вставкой
     --- 2740 ms --- время работы сортировки обменом
     ---  271 ms --- время работы сортировки Шелла
     ---  815 ms --- время работы турнирной сортировки
     ---  365 ms --- время работы быстрой сортировки

     ---   78 ms --- время работы сортировки с помощью встроенной функции